### PR TITLE
Add support for BeforeSendTransaction

### LINKED ISSUE #2077: Support BeforeSendTransaction
Per https://github.com/getsentry/team-mobile/issues/72, add a `BeforeSendTransaction` to work similarly as `BeforeSend` but for transactions instead of error events.

Complements transaction processors (#1468 / #1862).

### PR DESCRIPTION
**Pull request by SeanFeldman**
Add `BeforeSendTransaction` to work similarly to `BeforeSend` but for transactions instead of events.

Fixes #2077

### CODE REVIEW

File: src/Sentry/SentryClient.cs [hunk 1]
```diff
@@ -121,22 +121,69 @@ public void CaptureTransaction(Transaction transaction)
...
         {
             _options.ClientReportRecorder.RecordDiscardedEvent(DiscardReason.SampleRate, DataCategory.Transaction);
             _options.LogDebug("Transaction dropped by sampling.");
             return;
         }
 
+        var processedTransaction = BeforeSendTransaction(transaction);
+        if (processedTransaction is null) // Rejected transaction
+        {
+            // TODO: should the dropped transaction be recorded?
```
Comment (SeanFeldman) [inline src/Sentry/SentryClient.cs:138]: @mattjohnsonpint, should dropped transactions be recorded?

File: test/Sentry.Tests/SentryClientTests.verify.cs [hunk 2]
```diff
@@ -16,4 +16,18 @@ public Task CaptureEvent_BeforeEventThrows_ErrorToEventBreadcrumb()
 
         return Verify(@event.Breadcrumbs);
     }
+
+    [Fact]
+    public Task CaptureTransaction_BeforeSendTransactionThrows_ErrorToEventBreadcrumb()
```
Comment (SeanFeldman) [inline test/Sentry.Tests/SentryClientTests.verify.cs:21]: This is not working on my end. Is there any additional configuration required for Verify that I'm missing?

**Review by SeanFeldman — COMMENTED** [review]
While still a WIP, I have a few questions.

File: test/Sentry.Tests/SentryClientTests.verify.cs (hunk 2 above)
Comment (mattjohnsonpint) [reply test/Sentry.Tests/SentryClientTests.verify.cs:21]: There shouldn't be.  We did have an unrelated failure due to verify, but I've merged that in to this branch already.

File: src/Sentry/SentryClient.cs (hunk 1 above)
Comment (mattjohnsonpint) [reply src/Sentry/SentryClient.cs:138]: Yes.  Use `DiscardReason.BeforeSend` and `DataCategory.Transaction`.

File: src/Sentry/SentryClient.cs [hunk 3]
```diff
@@ -121,22 +121,69 @@ public void CaptureTransaction(Transaction transaction)
...
+            //_options.ClientReportRecorder.RecordDiscardedEvent(DiscardReason.BeforeSendTransaction, DataCategory.Error);
```
Comment (mattjohnsonpint) [inline src/Sentry/SentryClient.cs:139]:
```suggestion
            _options.ClientReportRecorder.RecordDiscardedEvent(DiscardReason.BeforeSend, DataCategory.Transaction);
```

File: test/Sentry.Tests/SentryClientTests.cs [hunk 4]
```diff
@@ -664,6 +664,77 @@ public void CaptureTransaction_DisposedClient_DoesNotThrow()
...
+
+        var sut = _fixture.GetSut();
+        sut.CaptureTransaction( new Transaction("test name", "test operation")
+        {
+            IsSampled = true,
+            EndTimestamp = DateTimeOffset.Now // finished
+        });
+
+        _fixture.ClientReportRecorder.Received(1)
+            .RecordDiscardedEvent(DiscardReason.BeforeSend, DataCategory.Error);
```
Comment (mattjohnsonpint) [inline test/Sentry.Tests/SentryClientTests.cs:735]:
```suggestion
            .RecordDiscardedEvent(DiscardReason.BeforeSend, DataCategory.Transaction);
```

File: test/Sentry.Tests/SentryClientTests.cs [hunk 5]
```diff
@@ -664,6 +664,77 @@ public void CaptureTransaction_DisposedClient_DoesNotThrow()
...
+
+        sut.CaptureTransaction(transaction);
+
+        Assert.Same(transaction, received);
+    }
+
+    [Fact(Skip = "TBD")]
+    public void CaptureTransaction_BeforeSendTransaction_RejectEvent_RecordsDiscard()
+    {
+        _fixture.SentryOptions.BeforeSend = _ => null;
```
Comment (mattjohnsonpint) [inline test/Sentry.Tests/SentryClientTests.cs:725]:
```suggestion
        _fixture.SentryOptions.BeforeSendTransaction = _ => null;
```

File: test/Sentry.Tests/SentryClientTests.verify.cs (hunk 2 above)
Comment (SeanFeldman) [reply test/Sentry.Tests/SentryClientTests.verify.cs:21]: PEBCAK. Fixing it 🙂

**Comment by mattjohnsonpint** [comment]
LGTM! 🚀

File: src/Sentry/SentryClient.cs [hunk 6]
```diff
@@ -121,22 +121,68 @@ public void CaptureTransaction(Transaction transaction)
...
 
+        var processedTransaction = BeforeSendTransaction(transaction);
+        if (processedTransaction is null) // Rejected transaction
+        {
+            _options.ClientReportRecorder.RecordDiscardedEvent(DiscardReason.BeforeSend, DataCategory.Transaction);
+            _options.LogInfo("Transaction dropped by BeforeSendTransaction callback.");
+            return;
+        }
+
         CaptureEnvelope(Envelope.FromTransaction(transaction));
```
Comment (vaind) [inline src/Sentry/SentryClient.cs:143]: I've just noticed this bit when working on Profiling - this should be sending `processedTransaction` instead

Comment (mattjohnsonpint) [reply src/Sentry/SentryClient.cs:143]: Yep!  @SeanFeldman - sorry I missed that in my review.   I think you can fix in a new PR.  Thanks.

Comment (SeanFeldman) [reply src/Sentry/SentryClient.cs:143]: Thank you for catching this, @vaind.
@mattjohnsonpint, I'll send a follow up PR.

Comment (SeanFeldman) [reply src/Sentry/SentryClient.cs:143]: Follow-up PR: #2204

**Pull request merged**